README.md
Klapperen
=========

<h3>Light weight text based chatbot by ultrazoid_</h3>
<br>
<b>Current Version:</b> (CODE) dev 1.6<br>
<b>Contact:</b> <a href="mailto:essper@ess.dudmail.com">Mail</a>
<h2>News</h2>
<ul>
<li>If there are any people who are looking at collaborating with me on this could they please contact me.</li>
<li>I am currently looking for people to create the tranlations of the responses once the std library has been made this will be in <code> *.csv </code> format</li>
<li>I am also looking for ideas for responses etc. to include in the library</li>
<li>The current English Library can be found <a href="http://ultrazoidserver.x10.mx/klapperen/library/ENlib.csv?dl">here.</a>
</ul>

<h2>Current Known Issues and Solutions:</h2>
<b>WOO!</b> There are currently no KNOWN issues. If you have an issue please post it under the issues tab.

<h2>Version List</h2>
For more info please check the <a href = "https://github.com/ultrazoid/Klapperen/blob/master/ChangeLog.md">Change Log</a><br>
<ul>
<li>(CODE) dev 1</li>
<ul>
<li>(CODE) dev 1.1</li>
<li>(CODE) dev 1.2</li>
<li>(CODE) dev 1.3</li>
<li>(CODE) dev 1.4</li>
<li>(CODE) dev 1.5</li>
<li>(CODE) dev 1.6</li>
</ul>
</ul>

ChangeLog.md
Change Log
==========
<ul>
<li>(CODE) dev 1</li>
<ul>
<li>Initial version</li>
</ul>
<li>(CODE) dev 1.1</li>
<ul>
<li>Fixed Issue #1</li>
</ul>
<li>(CODE) dev 1.2</li>
<ul>
<li>End code done</li>
<li>Version parsers</li>
<li>Fixed Issue 2</li>
</ul>
<li>(CODE) dev 1.3</li>
<ul>
<li>Added indicator to show the program worked (0fd44666f7)</li>
<li>Cleaned the output up with LINE BREAKS</li>
<li>Added update system for ENlib.csv</li>
</ul>
<li>(CODE) dev 1.4</li>
<ul>
<li>Bug fixes</li>
</ul>
<li>(CODE) dev 1.5</li>
<ul>
<li>Began working on actual chat section</li>
<li>Working saving log system</li>
</ul>
<li>(CODE) dev 1.6</li>
<ul>
<li>Moved stuff around</li>
<li>GRAMMAR!</li>
<li>Added catch for non supported language</li>
<li>Pre language select exit possible</li>
</ul>
</ul>

Main/english.py
'''
Klapperen Chatbot
Version (CODE) dev 1.6

English run-time for Klapperen Chatbot

Created on 30/08/2012

@author: ultrazoid_
'''
import string
import time
import os
import Functions
import random

def main(lfd, verP, verNu, info, verPN, k, s, y):
    print 'ENGLISH SELECTED'
    Functions.logMakeEN(lfd,verP,verNu)
    try:
        Functions.logUpd('EN',info)
        greetings = Functions.listGet(0,'EN')
        farewells = Functions.listGet(1,'EN')
        sGreetings = Functions.listGet(2,'EN')
        questions = Functions.listGet(3,'EN')
        stalls = Functions.listGet(4,'EN')
        sReplies = Functions.listGet(5, 'EN')
        ssReplies = Functions.listGet(6, 'EN')
        print "Library loaded!"
        raw_input("Press enter to continue...")
        lib = True
    except IOError:
        Functions.noLog('EN', 'ENGLISH')
        try:
            greetings = Functions.listGet(0,'EN')
            farewells = Functions.listGet(1,'EN')
            sGreetings = Functions.listGet(2,'EN')
            questions = Functions.listGet(3,'EN')
            stalls = Functions.listGet(4,'EN')
            sReplies = Functions.listGet(5, 'EN')
            ssReplies = Functions.listGet(6, 'EN')
            print "Library loaded!"
            raw_input("Press enter to continue...")
            lib = True
        except IOError:
            print 'FATAL ERROR ENCOUNTED'
            print 'PROGRAM WILL NOW SHUT DOWN'
            raw_input("Press enter to continue...")
    red = raw_input("Would you like to begin now(y/n)?")
    cont = False
    while cont == False:
        if red == 'y':
            print "Klapperen will now initiate"
            time.sleep(1)
            cont = True
        elif red == 'n':
            print "Klapperen will now exit"
            time.sleep(2)
            os.sys.exit()
        else:
            print 'Please enter a valid value'
            red = raw_input("Would you like to begin now(y/n)?")
    print "\nWelcome to Klapperen"
    print "Version",verPN
    print "By ultrazoid_"
    raw_input("Press enter to begin chat(Chat sequence not complete)")
    klp = k+"Hi my name is Klapperen:"
    en_inp = string.lower(raw_input(klp))
    en_wr = y+en_inp
    Functions.logWrite(lfd,klp)
    while en_inp != 'end':
        Functions.logWrite(lfd,en_wr)
        if en_inp in greetings:
            klp = k+greetings[random.randrange(0,len(greetings))]
            Functions.logWrite(lfd,klp)
            print klp
            klp = k+sGreetings[random.randrange(0,len(sGreetings))]
            Functions.logWrite(lfd,klp)
            print klp
        elif en_inp in farewells:
            klp = k+farewells[random.randrange(0,len(farewells))]
            Functions.logWrite(lfd,klp)
            print klp
            lin = s+"IF YOU WOULD LIKE TO EXIT PLEASE TYPE END"
            Functions.logWrite(lfd,lin)
            print lin
        elif en_inp in sGreetings:
            klp = k+sReplies[random.randrange(0,len(sReplies))]
            Functions.logWrite(lfd,klp)
            print klp
            klp = k+ssReplies[random.randrange(0,len(ssReplies))]
            Functions.logWrite(lfd,klp)
            print klp
        else:
            klp = k+stalls[random.randrange(0,len(stalls))]
            Functions.logWrite(lfd,klp)
            print klp
        en_inp = string.lower(raw_input('RESPONSE: '))
        en_wr = y+en_inp
    klp = k+farewells[random.randrange(0,len(farewells))]
    Functions.logWrite(lfd,klp)
    print klp
    print s+'ENDING'
    lin = s+'USER TERMINATED CONVERSTATION'
    Functions.logWrite(lfd,lin)
    print lin
    lin = 'LOG END'
    Functions.logWrite(lfd,lin)
    print s+'KLAPPEREN WILL NOW TERMINATE'
    raw_input('press enter to end...')

Main/Functions.py
'''
Klapperen Chatbot
Version (CODE) dev 1.6

Non language specific Functions

Created on 28/06/2012

@author: ultrazoid_
'''
import string
import os
import urllib
import time

def listGet(lineNumber, lang): #lineNumber starts at 0. So line 1 is 0, 2 1 and so on
    filename = lang+"lib.csv"
    fileOpen = open(filename, 'r')
    fileLines = fileOpen.readlines()
    listGot = string.split(fileLines[lineNumber],',')
    snoop = listGot[len(listGot)-1]
    if '\n' in snoop:
        listGot[-1] = listGot[-1].strip()
        listDone = listGot
    else:
        listDone = listGot
    fileOpen.close()
    return listDone

def logMakeEN(lfd, verP, verN):
    lfn = lfd+".txt"
    lfo = open(lfn , 'a')
    version = verP+' '+ verN
    lfo.write("KLAPPEREN LOG\n")
    lfo.write("Date: ")
    lfo.write(lfd+'\n')
    lfo.write("Version: ")
    lfo.write(version)
    lfo.write("\n")
    lfo.close()
    
def logWrite(logname, value):
    filename = logname+'.txt'
    logOpen = open(filename, 'a')
    logOpen.seek(2)
    logOpen.write(value+'\n')
    logOpen.close()

def verUpd(info):
    try:
        verNf = open(info["verIN"],'r')
        verN = verNf.readlines()
        verNf.close()
        info["verL"] = verN[1] 
        need = True
    except IOError:
        os.remove('VersionIN.txt')
        urllib.urlretrieve('http://ultrazoidserver.x10.mx/klapperen/version.txt',info["verIN"])
        verNf = open(info["verIN"],'r')
        verN = verNf.readlines()
        verNf.close()
        info["verL"] = verN[1] 
        need = False
    if need == True:
        urllib.urlretrieve('http://ultrazoidserver.x10.mx/klapperen/version.txt','VersionIN.txt')
        verINf = open('VersionIN.txt','r')
        verIN = verINf.readlines()
        verINf.close()
        info["verLN"] = verIN[1]

def noLog(lang, langL):
    print 'NO LIBRARY FOUND'
    dldInp = raw_input('WOULD YOU LIKE TO DOWNLOAD THE '+langL+' LIBRABRY(y/n)')
    cont = False
    while cont != True:
        if dldInp == 'y':
            print "Downloading"+lang+"lib.csv"
            urllib.urlretrieve('http://ultrazoidserver.x10.mx/klapperen/library/'+lang+'lib.csv',lang+'lib.csv')
            time.sleep(2)
            print 'Downloaded'                
            cont = True
        elif dldInp == 'n':
            print 'Program will now exit'
            print 'Please obtain a copy of the library'
            os.system('PAUSE')
            os.sys.exit()
        elif dldInp != 'y' or dldInp != 'n':
            print 'PLEASE ENTER A VALID RESPONSE.'
            dldInp = raw_input('WOULD YOU LIKE TO DOWNLOAD THE '+langL+ 'ONE(y/n)')

def logUpd(lang, info):
    print "Klapperen will now check to make sure "+lang+"lib.csv is up to date."
    if info["verLN"] != info["verL"]:
        print "There is an update for "+lang+"lib.csv."
        libUp = raw_input("update?:(y/n)")
        cont = False
        while cont != True:
            if libUp == 'y':
                os.remove(lang+'lib.csv')
                urllib.urlretrieve('http://ultrazoidserver.x10.mx/klapperen/library/'+lang+'lib.csv',lang+'lib.csv')
                os.remove('Version.txt')
                urllib.urlretrieve('http://ultrazoidserver.x10.mx/klapperen/version.txt',info["verIN"])
                print "UPDATE FINISHED"
                raw_input("Press enter to continue...")
                cont = True
            elif libUp == 'n':
                print 'Update aborted'
                print 'Program will now continue'
                raw_input("Press enter to continue...")
                cont = True
            elif libUp != 'y' or libUp != 'n':
                print "Please enter a valid response:"
    elif info["verLN"] == info["verL"]:
        print "No update needed "+lang+"lib.csv is up to date"
        raw_input("Press enter to continue...")

Main/Main.py
'''
Klapperen Chatbot
Version (CODE) dev 1.6

Main runtime

Created on 28/06/2012

@author: ultrazoid_

'''
import string
import Functions
import datetime
import english
import time

verP = '(CODE) dev'
verNu = '1.6'
verPN = verP+' '+verNu
info = {"Name":"Klapperen","Version":verPN,"verIN":"Version.txt","verL":"1","verLN":"1","Av":"EN"}
k = "KLP: "
y = "YOU: "
s = "SYS: "
#this is a random comment line
Functions.verUpd(info)

now = datetime.datetime.now()
lfd = now.strftime("%Y-%d-%m %H%M")

print info["Name"]
print info["Version"]
lanInt = raw_input('Please select a language(EN):')
LanInp = string.upper(lanInt)
cont = False
while cont == False:
    if LanInp == 'EN':
        cont = True
        english.main(lfd, verP, verNu, info, verPN, k, s, y)
        break
    elif LanInp == "END" or LanInp == "EXIT":
        print 'END/EXIT CODE RECOGNISED'
        time.sleep(1)
        break
    elif LanInp not in info["Av"]:
        cont = False
        print "Language unavailable"
        print "Currently only "+info["Av"]+" is available"
        print "Please try again!"
    lanInt = raw_input('Please select a language(EN):')
    LanInp = string.upper(lanInt)


        
                